File: leetcodeContest/188周/3.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {boolean[]} hasApple
 * @return {number}
 */
var minTime = function (n, edges, hasApple) {
  let fa = Array.fill(1e5 + 10).fill(0)
  let visit = Array.fill(1e5 + 10).fill(0)
  fa[0] = -1;
  for (let item of edges) {
    fa[item[1]] = item[0];
  }
  let ans = 0;
  for (let i = 0;i<hasApple.length;i++) {
    if (!i) {
      continue;
    }
    if (hasApple[i] && !visit[i]) {
      let node = i;
      visit[node] = 1;
      ans ++;
      while (fa[node] !== 0 && !visit[fa[node]]) {
        node = fa[node];
        visit[node] = 1;
        ans ++
      }
    }
  }
  return ans * 2;
}

File: leetcodeContest/188周/3.py
class Solution:
  # 把有苹果节点至根节点的所有节点加入集合set中，如果len(set)<=1时，结果为0，否则结果为（len(set)-1）*2
  def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
    ans = set()
    dic = {}
    for edge in edges:
      dic[edges[1]] = edge[0]
    
    for i in range(len(hasApple)):
      if hasApple[i] == True:
        ans.add(i)
        root = dic.get(i)
        while root is not None:
          ans.add(root)
          root = dic.get(root)
    n = len(ans)
    if n == 0 or n == 1:
      return 0
    else:
      return (n - 1) * 2